themes/default/content/blog/kubernetes-architecture-templates/index.md

```diff
@@ -0,0 +1,377 @@
+
+In a typical, very basic web application set up in a generic microservices architecture, you might have a data store, a backend, and a frontend (a user interface, or UI). In practice, you might have multiple backends and data stores to address multiple activities like logging in, running a shopping cart, delivering shipping updates, and more.
+
+To ensure all of these services work together, an orchestrator like Kubernetes ensures each service is running and healthy, is connected to the others, and runs as expected.
+
+## A typical Kubernetes architecture
+
+A Kubernetes _cluster_ is a collection of one or more machines (either virtual or bare-metal) called _nodes_ that host groups of _containers_, with each group containing multiple copies of a microservice. These groups of containers are called _pods_. And then, to connect and expose those pods to one another and the world, we add in _services_.
+
+For our system that we'll build today, we're going to have a cluster running on Google Kubernetes Engine (GKE). We'll create a namespace on that Kubernetes instance, which will hold all of our Kubernetes app resources. We'll have a _deployment_ that takes the configuration we need for NGINX (a _ConfigMap_) and spins up a pod holding the containers running that NGINX server. Then, we'll have a service that exposes the ports so we can access our NGINX server from anywhere.
+
+## Start with templates
+
+We're going to build our entire architecture starting with two templates: Our GKE template and our Helm app template.
+
+First, create a new directory to hold all of our project files, and add `app` and `infra` directories inside this root directory.
+
+```bash
+$ mkdir k8s-templates && cd k8s-templates
+$ mkdir app
+$ mkdir infra
+```
+
+In the `infra` directory, we'll add the GKE template:
+
+```bash
+$ cd infra
+$ pulumi new kubernetes-gcp-python
+```
+
+In the `app` directory, we'll add the web app template:
+
+```bash
+$ cd ../app
+$ pulumi new webapp-kubernetes-python
+```
+
+When it asks you for a project name in the `app` directory, give it the name `app-mi-k8s`.
+
+And now, the real fun begins!
+
+## Putting it all together
+
+We're going to use [Poetry](https://python-poetry.org/docs/) instead of the default virtual environment. In the root of the project, let's initialize a Poetry project. We'll need to define all of the dependencies from the application and infrastructure requirements:
+
+```bash
+$ cd ../ # if you need to go back to your project root
+$ poetry init
+# In the dynamic generation of dependencies, add pulumi, pulumi-gcp, and pulumi-kubernetes.
+```
+
+Now, we can pull the metadata and config files for Pulumi in:
+
+```bash
+$ touch Pulumi.yaml
+$ cat infra/Pulumi.yaml > Pulumi.yaml
+$ touch Pulumi.dev.yaml
+$ cat infra/Pulumi.dev.yaml app/Pulumi.dev.yaml > Pulumi.dev.yaml
+```
+
+And do a quick cleanup of the `Pulumi.dev.yaml` file:
```

Comment:
The reason these aren't separate projects is because we're trying to show a one-stack deployment of a Kubernetes cluster and application using the templates as a base. The request was to do something more than simply walking through templates, and this was what I came up with as the best way to show both templates and build something more interesting/different. The project takes the output from one template and dumps it into the other.